migration/process-migration/src/main/java/io/camunda/migration/process/MigrationRunner.java

```diff
@@ -7,12 +7,125 @@
+
+  private static final String ELASTICSEARCH = "elasticsearch";
+  private final Adapter adapter;
+  private final ProcessMigrationProperties properties;
+  private ScheduledFuture<?> countdownTask;
+
+  public MigrationRunner(
+      final ProcessMigrationProperties properties, final ConnectConfiguration connect) {
+    this.properties = properties;
+    adapter =
+        connect.getType().equals(ELASTICSEARCH)
+            ? new ElasticsearchAdapter(properties, connect)
+            : new OpensearchAdapter(properties, connect);
+  }
+
   @Override
-  public void run() {}
+  public void run() {
+    final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+    LOG.info("Process Migration started");
+    try {
+      String lastMigratedProcessDefinitionKey = adapter.readLastMigratedEntity();
+      List<ProcessEntity> items = adapter.nextBatch(lastMigratedProcessDefinitionKey);
+      while (shouldContinue(items)) {
+        if (!items.isEmpty()) {
+          lastMigratedProcessDefinitionKey = migrateBatch(items);
+        }
+        scheduleNextBatch(scheduler);
+        items = adapter.nextBatch(lastMigratedProcessDefinitionKey);
+      }
+    } catch (final Exception e) {
+      terminate(scheduler);
+      throw e;
+    }
+    terminate(scheduler);
+    LOG.info("Process Migration completed");
+  }
+
+  private boolean shouldContinue(final List<ProcessEntity> processes) {
+    if (!processes.isEmpty()) {
+      return true;
+    }
+    return countdownTask == null || !countdownTask.isDone();
+  }
+
+  private String migrateBatch(final List<ProcessEntity> processes) {
+    final String lastMigratedProcessDefinitionKey =
+        adapter.migrate(MigrationUtil.extractBatchData(processes));
+    adapter.writeLastMigratedEntity(lastMigratedProcessDefinitionKey);
+    return lastMigratedProcessDefinitionKey;
+  }
+
+  private void scheduleNextBatch(final ScheduledExecutorService scheduler) {
+    try {
+      if (countdownTask == null && isImporterFinished()) {
+        startCountdown(scheduler);
+      }
+      scheduler.schedule(() -> {}, adapter.getMinDelay().toSeconds(), TimeUnit.SECONDS).get();
+    } catch (final ExecutionException | InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      LOG.error("Schedule interrupted", ex);
+    }
+  }
+
+  private void startCountdown(final ScheduledExecutorService scheduler) {
+    LOG.info(
+        "Importer finished, migration will keep running for {}",
+        properties.getPostImporterTimeout());
+    countdownTask =
+        scheduler.schedule(
+            () -> {}, properties.getPostImporterTimeout().getSeconds(), TimeUnit.SECONDS);
+  }
+
+  private boolean isImporterFinished() {
+    final Set<ImportPositionEntity> importPositions;
+    try {
+      importPositions = adapter.readImportPosition();
+      return !importPositions.isEmpty()
+          && importPositions.stream().allMatch(ImportPositionEntity::getCompleted);
```

Comment:
Yes, they are being handled correctly. The adapter will always return a collection empty or without null items and the `getCompleted` is a primitive boolean so it would be _false_  if not set.